### PR TITLE
SYCL kernel Support for PReLU

### PR DESCRIPTION
**Pull request by AJAYRATNAM**
# Description

Introducing SYCL backend for the oneDNN PReLU primitive. This contains support for PReLU functionality in both forward and backward propagations.

# Supported scope
  Supported Data Types:
  Forward/ Backward : f32, s32, bf16, f16, s8, u8
 
## General

- [x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [x] Have you formatted the code using clang-format?

## Testing Scope
   `benchdnn: Passed`

Please refer the attached test output files.
[test_prelu_outputs.zip](https://github.com/oneapi-src/oneDNN/files/10912774/test_prelu_outputs.zip)

### CODE REVIEW

File: src/gpu/sycl/prelu_kernels.hpp [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_SYCL_PRELU_KERNELS_HPP
+#define GPU_SYCL_PRELU_KERNELS_HPP
+
+#include <cmath>
+#include <cstdlib>
+#include <stdlib.h>
+#include "CL/sycl/builtins.hpp"
```
Comment (densamoilov) [inline src/gpu/sycl/prelu_kernels.hpp:23]: What is this file used for? I don't see it in the specification. I think it's not intended for external use.

File: src/gpu/sycl/sycl_primitive_conf.hpp [hunk 2]
```diff
@@ -43,6 +46,70 @@ struct sycl_binary_conf_t {
...
+    int mask;
+    float sum;
+    broadcasting_strategy_t bcast_type;
+    int ndims;
+    int block_size;
+    int wg_size;
+    size_t n_thr;
+    size_t i_thr;
+};
+struct sycl_batch_normalization_conf_t {
```
Comment (densamoilov) [inline src/gpu/sycl/sycl_primitive_conf.hpp:72]: Irrelevant code, please remove

File: src/gpu/sycl/ref_prelu.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "gpu/sycl/ref_prelu.hpp"
+#include "common/utils.hpp"
+#include "gpu/sycl/prelu_kernels.hpp"
+
+#include <typeinfo>
+#include "gpu/nvidia/sycl_cuda_utils.hpp"
```
Comment (densamoilov) [inline src/gpu/sycl/ref_prelu.cpp:22]: SYCL kernels are backend agnostic therefore one cannot include backend specific headers here.

File: src/gpu/sycl/ref_prelu.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (densamoilov) [inline src/gpu/sycl/ref_prelu.cpp:2]: Please update copyrights -> 2023

File: src/gpu/sycl/ref_prelu.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+status_t ref_prelu_bwd_t::pd_t::init_conf() {
+    if (has_zero_dim_memory()) return status::success;
+    conf_ = sycl_prelu_conf_t();
+    conf_.data_md = sycl_md_t(src_md(0));
+    conf_.weights_md = sycl_md_t(weights_md(0));
+    conf_.diff_data_md = sycl_md_t(diff_src_md(0));
+    conf_.diff_weights_md = sycl_md_t(diff_weights_md(0));
+    conf_.diff_dst_md = sycl_md_t(diff_dst_md(0));
+    conf_.dst_md = sycl_md_t(dst_md(0));
```
Comment (densamoilov) [inline src/gpu/sycl/ref_prelu.cpp:93]: dst_md is not defined for backward pass, this causes an assert hit in debug mode

File: src/gpu/sycl/ref_prelu.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+status_t ref_prelu_bwd_t::execute_backward(const exec_ctx_t &ctx) const {
+    if (pd()->has_zero_dim_memory()) return status::success;
+
+    parallel_for(ctx, kernel_, [&](::sycl::handler &cgh) {
+        auto data = CTX_IN_SYCL_KERNEL_MEMORY(DNNL_ARG_SRC);
+        auto weights = CTX_IN_SYCL_KERNEL_MEMORY(DNNL_ARG_WEIGHTS);
+        auto diff_data = CTX_IN_SYCL_KERNEL_MEMORY(DNNL_ARG_DIFF_SRC);
+        auto diff_weights = CTX_IN_SYCL_KERNEL_MEMORY(DNNL_ARG_DIFF_WEIGHTS);
```
Comment (densamoilov) [inline src/gpu/sycl/ref_prelu.cpp:126]: Please check if you IN/OUT macros correctly, `diff_weights` is an OUT memory here. Please check other memory parameters.

File: src/gpu/sycl/ref_prelu.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        auto diff_dst = CTX_OUT_SYCL_KERNEL_MEMORY(DNNL_ARG_DIFF_DST);
+        auto dst = CTX_OUT_SYCL_KERNEL_MEMORY(DNNL_ARG_DST);
```
Comment (densamoilov) [inline src/gpu/sycl/ref_prelu.cpp:128]: dst is not provided for backward, please remove it

File: src/gpu/sycl/ref_prelu.hpp [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_SYCL_GENERIC_PRELU_HPP
+#define GPU_SYCL_GENERIC_PRELU_HPP
+
+#include "common/broadcast_strategy.hpp"
+#include "common/dnnl_thread.hpp"
```
Comment (densamoilov) [inline src/gpu/sycl/ref_prelu.hpp:21]: You don't need host threading utils for GPU

File: src/gpu/sycl/prelu_kernels.hpp [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+#include "common/c_types_map.hpp"
+#include "common/dnnl_thread.hpp"
+#include "common/dnnl_traits.hpp"
+#include "common/math_utils.hpp"
+#include "common/memory_storage.hpp"
+#include "common/utils.hpp"
+#include "cpu/simple_q10n.hpp"
```
Comment (densamoilov) [inline src/gpu/sycl/prelu_kernels.hpp:30]: you should not use CPU files here, there is `sycl_simple_q10n` file that has all the quantization utils for SYCL kernels.

File: src/gpu/nvidia/sycl_cuda_engine.cpp [hunk 10]
```diff
@@ -37,8 +37,8 @@
 #include "gpu/nvidia/sycl_cuda_engine.hpp"
 #include "gpu/nvidia/sycl_cuda_scoped_context.hpp"
 #include "gpu/nvidia/sycl_cuda_stream.hpp"
-
 #include "gpu/sycl/ref_binary.hpp"
+#include "gpu/sycl/ref_prelu.hpp"
```
Comment (densamoilov) [inline src/gpu/nvidia/sycl_cuda_engine.cpp:41]: (random spot) please add documentation for this primitive

File: src/gpu/sycl/prelu_kernels.hpp (hunk 1 above)
Comment (AJAYRATNAM) [reply src/gpu/sycl/prelu_kernels.hpp]: Yes, this file is unused and hence removed.

File: src/gpu/sycl/ref_prelu.hpp (hunk 8 above)
Comment (AJAYRATNAM) [reply src/gpu/sycl/ref_prelu.hpp]: Yes, Removed dnnl_thread utils.

File: src/gpu/sycl/sycl_primitive_conf.hpp [hunk 11]
```diff
@@ -20,6 +20,9 @@
 #include "gpu/sycl/sycl_post_ops.hpp"
 #include "gpu/sycl/sycl_types.hpp"
 
+#include "common/broadcast_strategy.hpp"
+#define MAX_NDIMS 6
```
[the remark this answers is not in the document]
Comment (AJAYRATNAM) [reply src/gpu/sycl/sycl_primitive_conf.hpp]: Removed.

File: src/gpu/sycl/ref_prelu.cpp (hunk 4 above)
Comment (saisrujana) [reply src/gpu/sycl/ref_prelu.cpp]: updated copyrights -> 2023 in al the files of PReLU, i.e., ref_prelu.cpp, ref_prelu.hpp, prelu_kernels.hpp and sycl_primitive_conf.hpp

File: src/gpu/sycl/ref_prelu.cpp (hunk 5 above)
Comment (saisrujana) [reply src/gpu/sycl/ref_prelu.cpp]: Removed.

File: src/gpu/sycl/ref_prelu.cpp (hunk 6 above)
Comment (saisrujana) [reply src/gpu/sycl/ref_prelu.cpp]: Checked and updated all the memory parameters. Also tests are success on the updated code.

File: src/gpu/sycl/prelu_kernels.hpp (hunk 9 above)
Comment (saisrujana) [reply src/gpu/sycl/prelu_kernels.hpp]: Removed the unused headers

File: src/gpu/nvidia/sycl_cuda_engine.cpp (hunk 10 above)
Comment (saisrujana) [reply src/gpu/nvidia/sycl_cuda_engine.cpp:41]: Added. Please check and let us know if anything is missed. Thanks!

**Review by densamoilov — APPROVED** [review]

**Pull request merged**